### ISSUE TITLE
Station list order sometimes incorrect

### ISSUE THREAD

**Issue by leskerr** [MorseKOB/PyKOB#197]
When a user is sending, their station ID should move to the bottom of their station list. The latest version of MKOB produces correct behavior for @leskerr and @AESilky, but not for @jchausler.

**Comment by AESilky** [comment]
A few observations from a combination of @jchausler, @leskerr and @AESilky ...
1. It works correctly if only the GUI 'Circuit Closer' button or 'ESC' is used to open/close the circuit before sending
2. Typing in the keyboard window opens the circuit as long as the 'code sender' box is enabled (maybe not a bad feature, but not planned)
3. Opening and closing using the key's closer works, but doesn't move the local sender to the bottom of the list on the local machine
4. What to do if you try to close with 'Circuit Closer', '+', or ESC and the physical key closer is open?

For #4, I think the physical closer has to take priority over all of the rest. If a physical key is connected and enabled and the physical key is open, then the circuit is open, and nothing you do in the GUI/keyboard should change it. Only when the physical key-closer is closed (or a physical key isn't connected) should the GUI functions matter. @leskerr do you agree?

**Comment by AESilky** [comment]
I am going to focus first on the physical key closer opening/closing and correctly moving the local sender to the bottom of the local list.

We can look at the other aspects once this is working as expected.

**Comment by leskerr** [comment]
> 4. What to do if you try to close with 'Circuit Closer', '+', or ESC and the physical key closer is open?
> 
> For item 4, I think the physical closer has to take priority over all of the rest. If a physical key is connected and enabled and the physical key is open, then the circuit is open, and nothing you do in the GUI/keyboard should change it. Only when the physical key-closer is closed (or a physical key isn't connected) should the GUI functions matter. @leskerr do you agree?

Not necessarily. If the user's key lacks a circuit closer, then the Circuit Closer checkbox is the only way to close the circuit. In fact, this was the original intent of the checkbox. (The Escape key is a convenient alternative to clicking on the checkbox.)

**Comment by AESilky** [comment]
@leskerr Got it. In fact, I know remember you telling me that before.

What I have found is that there are many places that manipulate the 'Circuit Closer' checkbox directly, as well as many places that call `set_local_loop_active`. However, only going through `kobmain.from_circuit_closer`, which is triggered by `kobactions.doCircuitCloser` correctly triggers the change in sender for the local station. Opening the key's closer or typing '~' in the keyboard calls `set_local_loop_active` directly rather than triggering the circuit closer mechanism through `kobactions.doCircuitCloser`. The ESC key does go through the `doCircuitCloser` mechanism, so it works correctly.

**Comment by AESilky** [comment]
Of course, opening the key's closer or typing '~' need to be seen as an OPEN rather than a toggle, just as closing the key's closer or typing '+' need to be seen as a CLOSE rather than a toggle. However, we need these (in addition to the toggle) go through a single entry point to make sure all of the functionality is consistent.

**Comment by jchausler** [comment]
Although with a loop circuit and the physical key open, the loop is "dead", with a sounder driver circuit, the key and sounder loops are separate.  In fact if one has a real sounder connected to a sounder driver circuit but no key, the sounder should still work even though the key connection is absent and thus the key loop open.  In 2.5 in "normal" mode if the program is started up with such a situation or with a key and the key open, the circuit closer box is automatically checked thus closing the circuit.  So the physical circuit closer on the key is not the sole arbiter of circuit open/close state.  How this should be handled in 4.0 probably needs to be discussed.  Thinking about it, I have to believe that the 2.5 approach seems to work real well but on more than one occasion I've had to tell folks over the phone whose set up isn't working to tell me the status of the circuit closer box and to have them correct it so that their system works.  This has happened with both loop and sounder driver configurations.  Sooner (or later) they begin to understand and I no longer get those phone calls, at least from those folks...73, Chris-----Original Message-----

**Comment by AESilky** [comment]
I believe I have fixed this. I also made a slight change in how the station list works. This makes it different from MKOB 2.5, but I like it better. The change is:
1. The current sender is at the top.
2. A line separates the current sender from the rest of the list.
3. The rest of the list is ordered by how long it has been since they last sent. The most recent sender is at the bottom, and it moves up from there.
4. Stations that are connected but have never sent a message are indented.
